### ISSUE TITLE
Need of some advice for investigation of low performance

### ISSUE THREAD

**Issue by freak82** [F-Stack/f-stack#507]
Hi there,

We did performance benchmarks of few scenarios. The tests were run on Ubuntu Server 18.04 KVM machine (Linux kernel 4.15.0-96-generic) with 2 cores (Intel(R) Xeon(R) Gold 6139 CPU @ 2.30GHz) and 32GB RAM. 10GBs NIC was attached in passthrough mode. The scenarios were:
1. Standard Linux kernel based transparent TCP proxy. The network card interrupts were set to run on CPU 1 via /proc/irq/<num>/smp_affinity_list. The proxy was single threaded and pinned to the same core (CPU 1) via cgroups. **The proxy was able to handle about 1.7-1.8 Gbps before started to throttle the traffic.**
2. DPDK/F-stack based transparent TCP proxy. It was set to run on the same CPU 1. The running config is given below. The DPDK setup were given 4096 huge pages x 2 MB. **The proxy was able to handle about 700-800 Mbps before started to throttle the traffic.**
**NOTE: The application layer code which handles the epoll and copy of the data between the sockets is absolutely the same for the proxy in point 1 and point 2.**
3. DPDK/F-stack based IP forward proxy. It was set to run on the same CPU 1. We did this test when trying to find out where the actual bottleneck is. In this test no sockets were open and the packets were simply forwared by the FreeBSD stack itself. **We redirected up to 5Gpbs for this test but we weren't able to reach the point were the traffic is throttled**.

Here is the `config.ini` for the DPDK/F-stack based proxy (without comments)
```
[dpdk]
lcore_mask=2
channel=4
promiscuous=0
numa_on=1
tx_csum_offoad_skip=0
tso=0
vlan_strip=0
idle_sleep=0
pkt_tx_delay=0
port_list=0
nb_vdev=0
nb_bond=0

[pcap]
enable = 0

[port0]
addr=XXX.XXX.XXX.252
netmask=XXX.XXX.XXX.248
broadcast=XXX.XXX.XXX.255
gateway=XXX.XXX.XXX.249

# pcap=/tmp/test.pcap

[kni]
enable=0

[freebsd.boot]
hz=100
fd_reserve=1024
kern.ncallout=524288
kern.sched.slice=1
kern.maxvnodes=524288
kern.ipc.nmbclusters=262144
kern.ipc.maxsockets=524000
net.inet.ip.fastforwarding=1
net.inet.tcp.syncache.hashsize=32768
net.inet.tcp.syncache.bucketlimit=32
net.inet.tcp.syncache.cachelimit=1048576
net.inet.tcp.tcbhashsize=524288
net.inet.tcp.syncache.rst_on_sock_fail=0
net.link.ifqmaxlen=4096
kern.features.inet6=0
net.inet6.ip6.auto_linklocal=0
net.inet6.ip6.accept_rtadv=2
net.inet6.icmp6.rediraccept=1
net.inet6.ip6.forwarding=0

[freebsd.sysctl]
kern.maxfiles=524288
kern.maxfilesperproc=524288
kern.ipc.soacceptqueue=4096
kern.ipc.somaxconn=4096
kern.ipc.maxsockbuf=16777216
kern.ipc.nmbclusters=262144
kern.ipc.maxsockets=524288
net.link.ether.inet.maxhold=5
net.inet.ip.redirect=0
net.inet.ip.forwarding=1
net.inet.ip.portrange.first=1025
net.inet.ip.portrange.last=65535
net.inet.ip.intr_queue_maxlen=4096
net.inet.tcp.syncache.rst_on_sock_fail=0
net.inet.tcp.rfc1323=1
net.inet.tcp.fast_finwait2_recycle=1
net.inet.tcp.sendspace=16384
net.inet.tcp.recvspace=16384
net.inet.tcp.cc.algorithm=cubic
net.inet.tcp.sendbuf_max=16777216
net.inet.tcp.recvbuf_max=16777216
net.inet.tcp.sendbuf_auto=1
net.inet.tcp.recvbuf_auto=1
net.inet.tcp.sendbuf_inc=16384
net.inet.tcp.recvbuf_inc=524288
net.inet.tcp.sack.enable=1
net.inet.tcp.msl=2000
net.inet.tcp.delayed_ack=1
net.inet.tcp.blackhole=2
net.inet.udp.blackhole=1
```

We also did additional test with about 200 Mbps traffic and with Linux perf enabled for the binary just to see where most of the time is spent. As far as I understand the results. They show that, aside from the packets processing loop, most of the time the application spent in the BSD functions for kqueue/event handling (I could be wrong thoug):
```
# Children      Self       Samples  Command          Shared Object       Symbol                                               
# ........  ........  ............  ...............  ..................  .....................................................
#
    43.46%    39.67%          9071  xproxy.release   xproxy.release      [.] main_loop
            |          
            |--35.31%--main_loop
            |          |          
            |           --3.71%--_recv_raw_pkts_vec_avx2
            |          
            |--5.44%--0x305f6e695f676e69
            |          main_loop
            |          
             --2.68%--0
                       main_loop

    25.51%     0.00%             0  xproxy.release   xproxy.release      [.] 0x0000000000cdbc40
            |
            ---0xcdbc40
               |          
               |--5.03%--__cap_rights_set
               |          
               |--4.65%--kern_kevent
               |          
               |--3.85%--kqueue_kevent
               |          
               |--3.62%--__cap_rights_init
               |          
               |--3.45%--kern_kevent_fp
               |          
               |--1.90%--fget
               |          
               |--1.61%--uma_zalloc_arg
               |          
                --1.40%--fget_unlocked

    10.01%     0.00%             0  xproxy.release   [unknown]           [k] 0x00007fa0761d8010
            |
            ---0x7fa0761d8010
               |          
               |--4.23%--ff_kevent_do_each
               |          
               |--2.33%--net::ff_epoll_reactor_impl::process_events <-- Only this function is ours
               |          
               |--1.96%--kern_kevent
               |          
                --1.48%--ff_epoll_wait

     7.13%     7.12%          1627  xproxy.release   xproxy.release      [.] kqueue_kevent
            |          
            |--3.84%--0xcdbc40
            |          kqueue_kevent
            |          
            |--2.41%--0
            |          kqueue_kevent
            |          
             --0.88%--kqueue_kevent

     6.82%     0.00%             0  xproxy.release   [unknown]           [.] 0x0000000001010010
            |
            ---0x1010010
               |          
               |--2.40%--uma_zalloc_arg
               |          
                --1.22%--uma_zero_item
```

So, from all of the tests we done and from the performance profiling it seems that the bottleneck lays somewhere in the FreeBSD event handling and/or in the TCP sockets handling.
However, I'm not sure at the moment if I'm not doing something wrong: probably the FreeBSD stack needs different settings or it has some significant differences from the Linux epoll and for example EPOLLET (which is emulated by F-stack) shouldn't be used?
I'm not sure at the moment why the performance of the proxy with the FreeBSD stack is so much below the performance of the standard Linux kernel proxy.
**Can you give me some advice where to look for some clue, or which settings to tweak, something which will help me improve the performance?**

Any help is appreciated!
Thanks,
Pavel.

**Comment by freak82** [comment]
Hi,

We managed to increase the performance to 900 Mbps - 1Gpbs using different ratio between the burst read packets and processed epoll events. However, it's still lower than the standard Linux kernel performance (1.7-1.8Gbps) for the same scenario of single threaded application.
We are aware that the F-stack application will scale linearly with the number of cores/processes while the Linux kernel one won't but we still would like to know if this performance on single core is epxected or not and if we can do anything to increase the performance further?

Thanks,
Pavel.

**Comment by jfb8856606** [comment]
Now, it's known that low performance while running F-Stack with few connections, especially in proxy mode. The possible reason for this include memory copy and MTU size, etc.

So you can run more connections/processes to increace the throughput.

And we have a plan to support new zero copy API, however, we can not promise complete time.
If you want to change MTU size, it need modify the code of F-Stack lib, such as finish `kni_change_mtu` function, etc.

**Comment by freak82** [comment]
Hi,

Thanks for the response.

This is not a test with few connections.This is real traffic from one internet service provider. We just switched few pools of the traffic to the test machine for the duration of the tests. The connections were about 10000 and more.

There are same number of memory copies, I think, in the standard Linux proxy too and yet it manages to have better trhoughput than the F-Stack one on single core. This is the part that bothers us because we can't understand why this is the case.

Pavel.

**Comment by jfb8856606** [comment]
You can enable `FF_USE_PAGE_ARRAY=1` in `lib/Makefile`, and test it again.
I have some ideas, but I need test it later.
And if you can show your test code is better.

**Comment by vincentmli** [comment]
I am just curious to see if FF_USE_PAGE_ARRAY=1 improves in my load test, but I got  compiling error:
<pre>
ff_memory.c:102:45: error: initialization makes integer from pointer without a cast [-Werror=int-conversion]
 static uint64_t             ff_page_start = NULL, ff_page_end = NULL;
                                             ^~~~
ff_memory.c:102:65: error: initialization makes integer from pointer without a cast [-Werror=int-conversion]
 static uint64_t             ff_page_start = NULL, ff_page_end = NULL;
                                                                 ^~~~
ff_memory.c: In function ‘ff_mmap_init’:
ff_memory.c:232:29: error: initialization makes integer from pointer without a cast [-Werror=int-conversion]
     uint64_t    virt_addr = NULL;
                             ^~~~
</pre>

**Comment by freak82** [comment]
I tried to do a test with it.
I fixed the above compilation errors (it needs to be set equal to 0 not NULL because it's not a pointer) and few more simple replacements.
The problem is that it crashed at runtime. It was something related to the memory allocations. I didn't save the backtrace. I plan to repeat the test at least to get the stacktrace but I've another task currently. I think I'll be able to do more testing next week.
